### PR TITLE
Les 68 'Check is entered userName is unique'

### PR DESCRIPTION
**Pull request by Kurilo-Pavel**

### CODE REVIEW

**Review by OxygenO22 — APPROVED** [review]

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,10 @@
...
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+    <div>
+     <UserName minSymbol={3} maxSymbol={256}/>
```
Comment (Terapool) [inline src/App.tsx:7]: it is better to use const for storing minSymbol and maxSymbol:

const USERNAME = {
MIN: 3,
MAX: 256,
}

UserName minSymbol={USERNAME.MIN} maxSymbol={USERNAME.MAX}

File: src/components/UserName.tsx [hunk 2]
```diff
@@ -1,39 +1,57 @@
...
   maxSymbol: number;
 }
-const UserName = ({minSymbol, maxSymbol}: UserNameProps):JSX.Element => {
-  const userNameRegex = new RegExp('^[A-Z\d]{' + minSymbol + ',' + maxSymbol + '}$', 'i');
+const UserName = ({minSymbol, maxSymbol}: UserNameProps): JSX.Element => {
+  const userNameRegex = new RegExp('^[A-ZА-ЯЁ \d]{' + minSymbol + ',' + maxSymbol + '}$', 'i');
   const [value, setValue] = useState('');
-  const [isError, setIsError] = useState(false);
   const [isBlur, setIsBlur] = useState(false);
+  const [addStyle, setAddStyle] = useState('');
```
Comment (Terapool) [inline src/components/UserName.tsx:15]: bad naming for addStyle - looks like function name.

It is better to call [extraStyle, setExtraStyle]

File: src/components/UserName.tsx [hunk 3]
```diff
@@ -1,39 +1,57 @@
...
+  const [error, setError] = useState('');
+
+  const dispatch = useAppDispatch();
+  const status = useAppSelector((state) => state.user.login);
 
   const fieldHandler = (e: React.FormEvent<HTMLInputElement>) => {
-    if (userNameRegex.test(e.currentTarget.value)) {
-      setIsError(false);
+    setValue(e.currentTarget.value);
```
Comment (Terapool) [inline src/components/UserName.tsx:22]: you use 'e.currentTarget.value' too many times.

create const: 

const {value} = e.currentTarget

File: src/components/UserName.tsx [hunk 4]
```diff
@@ -1,39 +1,57 @@
...
+const UserName = ({minSymbol, maxSymbol}: UserNameProps): JSX.Element => {
+  const userNameRegex = new RegExp('^[A-ZА-ЯЁ \d]{' + minSymbol + ',' + maxSymbol + '}$', 'i');
   const [value, setValue] = useState('');
-  const [isError, setIsError] = useState(false);
   const [isBlur, setIsBlur] = useState(false);
+  const [addStyle, setAddStyle] = useState('');
+  const [error, setError] = useState('');
+
+  const dispatch = useAppDispatch();
+  const status = useAppSelector((state) => state.user.login);
```
Comment (Terapool) [inline src/components/UserName.tsx:19]: I think 'status' is too abstract name. Use more meaningful, e.g. userStatus, authorizationStatus, requestStatus, etc.

File: src/components/UserName.tsx [hunk 5]
```diff
@@ -1,39 +1,57 @@
...
+    setIsBlur(true);
   }
 
   return (
     <div className='userName'>
       <label className='userNameLabel'>Username</label>
       <input type='text'
              required
-             className={'userNameInput ' + `${(isError && isBlur) ? 'errorInput' : ''}`}
+             className={`userNameInput ${(error && isBlur) ? addStyle : ''}`}
```
Comment (Terapool) [inline src/components/UserName.tsx:48]: It is a mess. Install https://www.npmjs.com/package/classnames and use it for className. See documentation for details

File: src/components/UserName.tsx [hunk 6]
```diff
@@ -1,39 +1,57 @@
...
              onBlur={blurHandle}/>
-      {(isError && isBlur) && <span className='error'>Invalid username entered</span>}
+      {(error && isBlur) && <span className='error'>{error}</span>}
     </div>
   )
 }
-export default UserName;
\ No newline at end of file
+export default UserName;
+// violettakolendovich@gmail.com
```
Comment (Terapool) [inline src/components/UserName.tsx:56]: Remove comments

File: src/components/userName.scss [hunk 7]
```diff
@@ -26,4 +26,7 @@
   color: red;
   display: block;
   margin-top: 0.5em;
+}
+.redBorder{
```
Comment (Terapool) [inline src/components/userName.scss:30]: .redBorder{ - set a space before {

File: src/constants.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export const BACKEND_URL : string  =  'https://lessoner.herokuapp.com/en';
```
Comment (Terapool) [inline src/constants.ts:1]: no need to set string type for string const. Typescript makes type inference

File: src/index.tsx [hunk 9]
```diff
@@ -3,13 +3,17 @@ import ReactDOM from 'react-dom/client';
...
+import {Provider} from "react-redux";
+import store from "./store"
 
 const root = ReactDOM.createRoot(
   document.getElementById('root') as HTMLElement
 );
 root.render(
   <React.StrictMode>
+    <Provider store={store}>
     <App />
```
Comment (Terapool) [inline src/index.tsx:15]: No indent for App component - add tabulation as App component is nested into Provider

File: src/store/loginName/userSlice.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import {createAsyncThunk, createSlice, PayloadAction} from "@reduxjs/toolkit";
+import {BACKEND_URL} from "../../constants";
+
+export const getUser = createAsyncThunk(
+  'user/getUserStatus',
+  async (userName: string) => {
+    const response = await fetch(`${BACKEND_URL}/check_username?name=${userName}`);
+    const data = await response.json();
+    if (response.status === 200) {
```
Comment (Terapool) [inline src/store/loginName/userSlice.ts:9]: please empty line before if

File: src/store/loginName/userSlice.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      return data.usernameExists;
+    } else {
+      return null;
+    }
+  }
+);
+
+type User = {
+  login: boolean|any
```
Comment (Terapool) [inline src/store/loginName/userSlice.ts:18]: what??? boolean|any ???

File: src/components/UserName.tsx [hunk 12]
```diff
@@ -1,39 +1,57 @@
...
+
+  const dispatch = useAppDispatch();
+  const status = useAppSelector((state) => state.user.login);
 
   const fieldHandler = (e: React.FormEvent<HTMLInputElement>) => {
-    if (userNameRegex.test(e.currentTarget.value)) {
-      setIsError(false);
+    setValue(e.currentTarget.value);
+    dispatch(getUser(e.currentTarget.value));
+    if (!userNameRegex.test(e.currentTarget.value)) {
```
Comment (Terapool) [inline src/components/UserName.tsx:24]: please insert empty line before if

File: src/store/loginName/userSlice.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+}
+const initialState: User = {
+  login: false,
```
Comment (Terapool) [inline src/store/loginName/userSlice.ts:21]: please for boolean type use names with 'is', 'has' prefixes:

isLogged

**Review by NatashaMikhaltsova — APPROVED** [review]

File: src/components/UserName.tsx [hunk 14]
```diff
@@ -1,38 +1,57 @@
...
   }
 
+  useEffect(() => {
+    if (userStatus) {
+      setError('User already exists. Please enter a different username');
+      setExtraStyle('redBorder')
+    }
+  }, [userStatus]);
+
   const blurHandle = (e: React.FormEvent<HTMLInputElement>) => {
```
Comment (Terapool) [inline src/components/UserName.tsx:42]: const blurHandle = (e: React.FormEvent<HTMLInputElement>) => {
 
 Remove e from arguments. you do not use it

**Review by Terapool — APPROVED** [review]

**Comment by Evgeniy121212** [comment]
@Kurilo-Pavel , please, update the branch with new changes

**Pull request merged**